### ISSUE TITLE
Replace `IsMatrix => IsMatrixObj` implication, by introducing `IsMatrixOrMatrixObj`?

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#4503]
(This issue is closely related to issue #2972)

Currently, we have an implication `IsMatrix => IsMatrixObj`:
```gap
InstallTrueMethod( IsMatrixObj, IsMatrix );
```
The motivation for that was to make it easier to write code that works both for old-style list-of-list matrices, and new-style MatrixObj matrices.

The drawback is that it is harder to write code which *requires* new style MatrixObj, e.g. to support empty matrices, or to get accurate `BaseDomain` results.

As a compromise, we could instead introduce a new filter/category/whatever `IsMatrixOrMatrixObj` in analogy to `IsListOrCollection`:
```gap
DeclareCategory( "IsMatrixOrMatrixObj", IsObject );
InstallTrueMethod( IsMatrixOrMatrixObj, IsMatrix );
InstallTrueMethod( IsMatrixOrMatrixObj, IsMatrixObj );
```
The drawback would be that we'd have to adjust a lot of operations and methods to use this filter. But we already have to do that anyway, the question is just whether we change to `IsMatrixOrMatrixObj` or to `IsMatrix`. On the upside, we can also start to get rid of some hacks; e.g. 
```gap
InstallMethod( ZeroMutable,
    [ IsMatrixObj ],
    function( M )
    if IsPlistRep( M ) then
      TryNextMethod();
    else
      return Matrix( ZeroMutable( Unpack( M ) ), M );
    fi;
    end );
```
could become
```gap
InstallMethod( ZeroMutable,
    [ IsMatrixObj ],
    M -> Matrix( ZeroMutable( Unpack( M ) ), M ) );
```
(for that function one may then also wonder whether perhaps using `ZeroMatrix` or `NewZeroMatrix` would be more appropriate, but that's a separate concern).

Are there perhaps other negative consequences of this that I am overlooking? @ThomasBreuer might be able to think of some...

**Comment by ThomasBreuer** [comment]
The question is about generic methods (and about the documentation of the MatrixObj setup).

I think the main reason why on wants to have different generic methods for an `IsPlistRep` matrix and a non-`IsPlistRep` MatrixObj is that the code for the non-`IsPlistRep` MatrixObj delegates to `IsPlistRep` objects (via `Unpack`), and using this code for an `IsPlistRep` matrix would mean infinite recursion.
This is how I interpret the `ZeroMutable` method shown above.
One could argue that a method for `IsMatrix` is anyhow available and should have higher rank, and thus the `TryNextMethod` call in case of `IsPlistRep` will not be reached. In this sense, one could simplify the methods in question already in the current setup.
(Perhaps an `Assert` about non-`IsPlistRep` could be added.)

Are there other situations where a filter describing "`IsMatrixObj` and not `IsPlistRep`" would be useful?

**Comment by fingolfin** [comment]
We just discussed this, and I explained that I see some more reasons for wanting this: some people (e.g. the semigroups package team) really would like to write code that *requires* MatrixObj everywhere, and deliberately does not try to support old-style matrices, so that they can take full advantage of things like NrCols, BaseDomain etc., without having to worry about inputs where these run into errors. Also, recently I tried to adapt a bunch of things in the GAP library to work better with MatrixObj, and I had to write glue code to support old-style matrices a lot; it would be useful if I could easily distinguish between the two kinds of matrices... OK, I guess I could write something like `IsMatrixObj(x) and not IsMatrix(x)`,  but I can't use this as filter for a method installation.

**Comment by ssiccha** [comment]
Would it then make sense to also have a `IsVectorOrIsVectorObj` or an `IsRowVectorOrIsVectorObj`?

**Comment by ThomasBreuer** [comment]
@ssiccha This question is raised in `lib/matobj1.gd`. Up to now, I have regarded it as not interesting because the viewpoint of a "common roof" for `IsRowVector` and `IsVectorObj` would be useful only in a few places. After the introduction of `IsMatrixOrMatrixObj`, I must admit that the code looks quite asymmetric w.r.t. vector and matrix objects, but this alone is not convincing.

**Comment by fingolfin** [comment]
Resolved by PR #4518
